File: proxy/test/llm-client.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { QwenClient } from '../src/llm-client';

describe('QwenClient.streamChat', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
  });

  it('streams chunks and parses SSE data lines', async () => {
    const sseBody = [
      'data: {"choices":[{"delta":{"content":"你"}}]}\n',
      'data: {"choices":[{"delta":{"content":"好"}}]}\n',
      'data: [DONE]\n',
    ].join('');

    const mockFetch = vi.fn().mockResolvedValue(
      new Response(sseBody, { status: 200, headers: { 'content-type': 'text/event-stream' } }),
    );
    vi.stubGlobal('fetch', mockFetch);

    const client = new QwenClient('test-key', 'qwen-plus', 'https://example.com/v1/chat/completions');
    const chunks: string[] = [];
    for await (const chunk of client.streamChat({
      messages: [{ role: 'user', content: 'hi' }],
      temperature: 0.7,
    })) {
      chunks.push(chunk);
    }
    expect(chunks).toEqual(['你', '好']);
  });

  it('throws on non-200 response', async () => {
    vi.stubGlobal('fetch', vi.fn().mockResolvedValue(new Response('bad', { status: 500 })));
    const client = new QwenClient('k', 'm', 'https://e');
    await expect(async () => {
      for await (const _ of client.streamChat({ messages: [], temperature: 0 })) { /* noop */ }
    }).rejects.toThrow();
  });
});

File: app/test/data/llm/llm_client_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:gym_coach/data/llm/llm_client.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

void main() {
  test('client sends signed POST and parses SSE', () async {
    final mock = MockClient((req) async {
      expect(req.headers['X-Signature'], isNotEmpty);
      expect(req.headers['X-Timestamp'], isNotEmpty);
      expect(req.body, contains('"messages"'));
      return http.Response(
        'data: {"delta":"好"}\n\ndata: [DONE]\n\n',
        200,
        headers: {'content-type': 'text/event-stream'},
        encoding: utf8,
      );
    });

    final client = LlmClient(
      endpoint: 'https://example.com',
      hmacSecret: 'secret',
      httpClient: mock,
    );
    final chunks = await client.chat(
      messages: [
        {'role': 'user', 'content': 'hi'},
      ],
    ).toList();
    expect(chunks, ['好']);
  });

  test('retries on 500 and eventually succeeds', () async {
    var attempts = 0;
    final mock = MockClient((req) async {
      attempts++;
      if (attempts < 2) {
        return http.Response('boom', 500);
      }
      return http.Response(
        'data: {"delta":"ok"}\n\ndata: [DONE]\n\n',
        200,
        headers: {'content-type': 'text/event-stream'},
      );
    });
    final client = LlmClient(
      endpoint: 'https://example.com',
      hmacSecret: 's',
      httpClient: mock,
    );
    final chunks = await client.chat(messages: [
      {'role': 'user', 'content': 'hi'},
    ]).toList();
    expect(chunks, ['ok']);
    expect(attempts, 2);
  });
}